weave/flow/dataset.py

```diff
@@ -236,3 +254,169 @@ def select(self, indices: Iterable[int]) -> Self:
+            Exception: Any exception raised by the function during processing is propagated.
+
+        Examples:
+            ```python
+            import weave
+
+            # Create a dataset
+            ds = weave.Dataset(rows=[
+                {"id": 1, "text": "Hello world", "lang": "en"},
+                {"id": 2, "text": "Bonjour monde", "lang": "fr"},
+                {"id": 3, "text": "Hola mundo", "lang": "es"}
+            ])
+
+            # Example 1: Simple transformation with a single parameter
+            def add_text_length(text):
+                return {"length": len(text)}
+
+            # Example 2: Multi-parameter function with type hints
+            def combine_fields(id: int, text: str, lang: str) -> dict:
+                return {
+                    "display": f"[{id}] {text} ({lang})",
+                    "uppercase": text.upper()
+                }
+
+            # Example 3: Function with default parameters
+            def classify_language(lang, default_category="other"):
+                categories = {"en": "english", "es": "spanish", "fr": "french"}
+                return {"category": categories.get(lang, default_category)}
+
+            # Apply the transformations
+            ds_with_length = ds.map(add_text_length)
+            ds_combined = ds.map(combine_fields)
+            ds_categorized = ds.map(classify_language)
+
+            # Example output for ds_with_length[0]:
+            # {
+            #     "id": 1,
+            #     "text": "Hello world",
+            #     "lang": "en",
+            #     "length": 11
+            # }
+            ```
+        """
+        processed_rows: list[tuple[int, dict]] = []
+
+        # Inspect the function signature
+        sig = inspect.signature(func)
+        param_names = list(sig.parameters.keys())
+
+        # Convert lambda to named function for better async handling
+        func = wrap_lambda(func)
+
+        # Store the function name before applying weave.op
+        func_name = getattr(func, "__name__", "unknown_function")
+
+        # Apply weave.op after lambda conversion
+        func = weave.op(func)
+
+        async def process_row(row: dict) -> dict:
+            try:
+                # Extract and pass specific parameters
+                kwargs = {}
+                for name in param_names:
+                    if name in row:
+                        kwargs[name] = row[name]
+                    else:
+                        # Skip this parameter or use a default if available
+                        if sig.parameters[name].default is not inspect.Parameter.empty:
+                            kwargs[name] = sig.parameters[name].default
+                        else:
+                            raise ValueError(
+                                f"Function expects parameter '{name}' but this column "
+                                f"is not in the dataset row: {short_str(row)}"
+                            )
```

Comment:
Is this basically saying:

"Select the subset of row keys that are relevant for this func"?

Reply:
yeap